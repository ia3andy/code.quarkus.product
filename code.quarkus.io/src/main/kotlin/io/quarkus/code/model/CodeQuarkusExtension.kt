package io.quarkus.code.model

data class CodeQuarkusExtension(
        val id: String,
        val shortId: String,
        val version: String,
        val name: String,
        val description: String?,
        val shortName: String?,
        val category: String,
        val tags: List<String>,
        val keywords: List<String>,
        val providesExampleCode: Boolean,
        val guide: String?,
        val order: Int,

        @Deprecated(message = "no continued")
        val default: Boolean,

        @Deprecated(message = "has been replaced", replaceWith = ReplaceWith("tags"))
        val status: String,

        @Deprecated(message = "has been replaced", replaceWith = ReplaceWith("keywords"))
        val labels: List<String>
)