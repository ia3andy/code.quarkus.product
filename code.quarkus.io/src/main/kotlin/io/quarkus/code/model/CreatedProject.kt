package io.quarkus.code.model

data class CreatedProject(
        var path: String
)