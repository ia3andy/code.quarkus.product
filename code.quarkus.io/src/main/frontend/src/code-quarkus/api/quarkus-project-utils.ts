import { parse, stringify } from 'querystring';
import { ExtensionEntry } from '../pickers/extensions-picker';
import { createGitHubProject } from './code-quarkus-github-api';
import { QuarkusProject } from './model';
import { BACKEND_URL, CLIENT_NAME } from './env';

export enum Target {
  DOWNLOAD = 'DOWNLOAD',
  SHARE = 'SHARE',
  GITHUB = 'GITHUB'
}

export interface GenerateResult {
  target: Target;
  url: string;
  shareUrl?: string;
}

export function generateProjectQuery(project: QuarkusProject, github: boolean = false): string {
  const params: any = {
    ...(project.metadata.groupId && { g: project.metadata.groupId }),
    ...(project.metadata.artifactId && { a: project.metadata.artifactId }),
    ...(project.metadata.version && { v: project.metadata.version }),
    ...(project.metadata.buildTool && { b: project.metadata.buildTool }),
    ...(project.metadata.noExamples && { ne: project.metadata.noExamples }),
    ...(project.extensions && { s: project.extensions.map(e => e.shortId).join('.') }),
    cn: CLIENT_NAME
  };
  if (github) {
    params.github = true;
  }
  return stringify(params);
}

const BASE_LOCATION = window.location.href.replace(window.location.search, '');

export function getProjectDownloadUrl(project: QuarkusProject) {
  const baseUrl = BACKEND_URL.startsWith('http') ? BACKEND_URL : BASE_LOCATION;
  return `${baseUrl.replace(/\/$/, '')}/d?${generateProjectQuery(project)}`;
}

export function getProjectShareUrl(project: QuarkusProject, github = false) {
  return `${BASE_LOCATION}?${generateProjectQuery(project, github)}`;
}

export async function generateProject(environment: string, project: QuarkusProject, target: Target): Promise<GenerateResult> {
  switch (target) {
    case Target.DOWNLOAD:
    case Target.SHARE:
      const url = getProjectDownloadUrl(project);
      const shareUrl = getProjectShareUrl(project);
      if (target !== Target.SHARE && environment !== 'dev') {
        setTimeout(() => window.open(url, '_blank'), 500);
      }
      return { target, url, shareUrl };
    case Target.GITHUB:
      const result = await createGitHubProject(project);
      return { target, url: result.url };
  }
}

export const createOnGitHub = (project: QuarkusProject, clientId: string) => {
  const authParams = {
    redirect_uri: getProjectShareUrl(project, true),
    client_id: clientId,
    scope: 'public_repo,workflow',
    state: Math.random().toString(36)
  };
  const githubAuthorizeUrl = `https://github.com/login/oauth/authorize?${stringify(authParams)}`;
  window.location.href = githubAuthorizeUrl;
};

export function newDefaultProject(): QuarkusProject {
  return ({
    metadata: {
      groupId: 'org.acme',
      artifactId: 'code-with-quarkus',
      version: '1.0.0-SNAPSHOT',
      buildTool: 'MAVEN',
      noExamples: false
    },
    extensions: [],
  });
}

const queryName = 'extension-search';

export function resolveInitialFilterQueryParam(): string {
  return getParams(queryName) || '';
}

const getParams = (paramName: string): string | null => {
  const searchParams = new URLSearchParams(window.location.search);
  return searchParams.get(paramName);
}

export function resolveInitialProject(extensions: ExtensionEntry[]) {
  return parseProjectInQuery(extensions) || newDefaultProject();
}

const defaultCleanHistory = () => {
  console.log(`remove query from url: ${window.location.search}`);
  window.history.replaceState({}, document.title, window.location.href.replace(window.location.search, ''));
};

export function parseProjectInQuery(extensions: ExtensionEntry[],
                                    search: string = window.location.search.substr(1),
                                    cleanHistory: () => void = defaultCleanHistory): QuarkusProject | undefined {
  if (search.length === 0) {
    return undefined;
  }
  const queryObj = parse(search);
  const shortIds = new Set((typeof queryObj.s === 'string') ? (queryObj.s as string).split('.') : []);
  const ids = new Set();
  const selectedExtensions = extensions.filter(e => {
    if (shortIds.has(e.shortId)) {
      const already = ids.has(e.id);
      ids.add(e.id);
      return !already;
    }
    return false;
  });
  const defaultProj = newDefaultProject();
  const project = {
    metadata: {
      groupId: queryObj.g || defaultProj.metadata.groupId,
      artifactId: queryObj.a || defaultProj.metadata.artifactId,
      version: queryObj.v || defaultProj.metadata.version,
      buildTool: queryObj.b || defaultProj.metadata.buildTool,
      noExamples: queryObj.ne || defaultProj.metadata.noExamples
    },
    extensions: selectedExtensions,
    github: queryObj.github === 'true' ? {
      code: queryObj.code,
      state: queryObj.state
    } : undefined
  } as QuarkusProject;
  if (project.github) {
    console.log('Received GitHub auth');
  }
  cleanHistory();
  return project;
}