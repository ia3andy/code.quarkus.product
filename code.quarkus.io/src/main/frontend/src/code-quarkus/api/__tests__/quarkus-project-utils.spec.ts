import { cleanup } from '@testing-library/react';
import { parseProjectInQuery } from '../quarkus-project-utils';
import { Extension } from '../model';
import {parse} from "querystring";

afterEach(() => {
  cleanup();
});

const entries: Extension[] = [
  {
    'id': 'io.quarkus:quarkus-arc',
    'name': 'ArC',
    'version': 'test-version',
    'shortId': 'a',
    'tags': [],
    'keywords': [
      'arc',
      'cdi',
      'dependency-injection',
      'di',
      'label'
    ],
    'default': false,
    'description': 'Build time CDI dependency injection',
    'shortName': 'CDI',
    'category': 'Core',
    'order': 0,
  },
  {
    'id': 'io.quarkus:quarkus-camel-netty4-http',
    'version': 'test-version',
    'name': 'Camel Netty4 test HTTP',
    'tags': ['preview'],
    'shortId': 'b',
    'default': false,
    'keywords': [
      'camel-netty4-http',
      'camel'
    ],
    'description': 'Camel support for Netty',
    'category': 'Integration',
    'order': 2,
  },
  {
    'id': 'some-id',
    'shortId': 'c',
    'version': 'test-version',
    'name': 'A CDI in name test',
    'tags': ['experimental'],
    'default': false,
    'keywords': [
      'lambda',
      'amazon-lambda',
      'aws-lambda',
      'amazon',
      'aws',
      'label'
    ],
    'shortName': 'a shortname',
    'description': 'Some description',
    'category': 'Cloud',
    'order': 1,
  },
  {
    'id': 'some-id',
    'shortId': 'c',
    'version': 'test-version',
    'name': 'A CDI in name test',
    'tags': ['experimental'],
    'default': false,
    'keywords': [
      'lambda',
      'amazon-lambda',
      'aws-lambda',
      'amazon',
      'aws',
      'label'
    ],
    'shortName': 'a shortname',
    'description': 'Some description',
    'category': 'Toto',
    'order': 1,
  },
];

describe('quarkus-project', () => {
  it('parseProjectInQuery correctly', () => {
    const queryParams = parse('g=org.test&a=code-test&v=1.0.0-SNAPSHOT&b=MAVEN&c=org.toto&s=a.bG');
    const parsedProject = parseProjectInQuery(entries, queryParams);
    expect(parsedProject).toMatchSnapshot('no-github');
  });

  it('parseProjectInQuery correctly with github', () => {
    const queryParams = parse('g=org.test&a=code-test&v=1.0.0-SNAPSHOT&b=GRADLE&c=org.toto&s=a.bG&code=totototo&state=djdjdj&github=true');
    const parsedProject = parseProjectInQuery(entries, queryParams);
    expect(parsedProject).toMatchSnapshot('github');
  });

  it('parseProjectInQuery with some other metadata', () => {
    const parsedProject = parseProjectInQuery(entries, parse('v=1.0'));
    expect(parsedProject).toMatchSnapshot('v=1.0');
  });

});