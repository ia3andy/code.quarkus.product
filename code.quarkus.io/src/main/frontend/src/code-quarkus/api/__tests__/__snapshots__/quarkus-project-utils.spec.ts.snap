// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`quarkus-project parseProjectInQuery correctly with github: github 1`] = `
Object {
  "extensions": Array [
    Object {
      "category": "Core",
      "default": false,
      "description": "Build time CDI dependency injection",
      "id": "io.quarkus:quarkus-arc",
      "keywords": Array [
        "arc",
        "cdi",
        "dependency-injection",
        "di",
        "label",
      ],
      "name": "ArC",
      "order": 0,
      "shortId": "a",
      "shortName": "CDI",
      "tags": Array [],
      "version": "test-version",
    },
  ],
  "github": Object {
    "code": "totototo",
    "state": "djdjdj",
  },
  "metadata": Object {
    "artifactId": "code-test",
    "buildTool": "GRADLE",
    "groupId": "org.test",
    "noExamples": false,
    "version": "1.0.0-SNAPSHOT",
  },
}
`;

exports[`quarkus-project parseProjectInQuery correctly: no-github 1`] = `
Object {
  "extensions": Array [
    Object {
      "category": "Core",
      "default": false,
      "description": "Build time CDI dependency injection",
      "id": "io.quarkus:quarkus-arc",
      "keywords": Array [
        "arc",
        "cdi",
        "dependency-injection",
        "di",
        "label",
      ],
      "name": "ArC",
      "order": 0,
      "shortId": "a",
      "shortName": "CDI",
      "tags": Array [],
      "version": "test-version",
    },
  ],
  "github": undefined,
  "metadata": Object {
    "artifactId": "code-test",
    "buildTool": "MAVEN",
    "groupId": "org.test",
    "noExamples": false,
    "version": "1.0.0-SNAPSHOT",
  },
}
`;

exports[`quarkus-project parseProjectInQuery with some other metadata: v=1.0 1`] = `
Object {
  "extensions": Array [],
  "github": undefined,
  "metadata": Object {
    "artifactId": "code-with-quarkus",
    "buildTool": "MAVEN",
    "groupId": "org.acme",
    "noExamples": false,
    "version": "1.0",
  },
}
`;
